Add accelerator records and summaries (modifications for CASO and formatting)

nick-whyatt opened the pull request:
Editing the above PR to work with CASO by changing some string formatting and tests.

pjcon committed:
Allow RecordFactory to parse JSON messages using jsonschema

pjcon committed:
Add record factory json tests using accelerator record type

pjcon committed:
Add accelerator record type

pjcon committed:
Add accelerator summary type

Commit:
[GT-194] Add accelerator summaries and records

nick-whyatt committed:
Solve issues with capitalisation in message headers

- Styilistically, "Accelerator" message hurt our eyes
- Practically, cASO's accounting script is configured to set things to "APEL-accounting-message"
- This leads to autorejection of otherwise valid messages
- This changes the schema, and a bunch of tests to check if it works

nick-whyatt committed:
Added AcceleratorModels, added VAcceleratorSummaries, and time dependent joins

- AcceleratorModels table to add categories to specific cards
- VAcceleratorSummaries with time dependent join logic - nested subquery, VERY expensive
	- find a way to fix this
- Changes to various functions
- added GetNewModels procedure that gives starting rules for AcceleratorModels logic
- added SiteNameLookup
- various bug fixes

nick-whyatt committed:
Measurement year/month changes

nick-whyatt committed:
Progress commit to VAcceleratorSummaries: triggers, GetModelSummaries and structure

- Alters the structure of VAcceleratorSummaries to simply draw from AcceleratorModelSummaries
- This is to prevent researching the expensive but periodic subquery for time-dependent accounting
- Adds AcceleratorModelSummaries to insert things into
- Adds GetModelSummaries which is basically a REPLACE operation of the original VAcceleratorSummaries
into AcceleratorModelSummaries
- Adds an INSERT trigger (REPLACE triggers INSERT, or DELETE/INSERT) to call GetNewModels and then
GetModelSummaries

nick-whyatt committed:
Change of approach: moving to events over triggers

- Modified GetModelSummaries proven to work, tested end to end
- Triggers removed as MySQL ONLY SUPPORTS row-by-row triggers and the operation
was only designed as a statement-level trigger
- Moving to an event based system

nick-whyatt committed:
Add event model_summaries_hourly that hourly updates AcceleratorModelSummaries

- The query used to time-order is very expensive
- Having it in the view query is a nightmare
- Event updates another table every hour to have the right data

nick-whyatt committed:
Fixed jobs!

Jobs now actually run.  0.019s recall on VAcceleratorSummaries.

nick-whyatt committed:
Final schema alterations for GPU

- Combined everything into Accelerator.sql that adds events, tables and methods
- Added vital (previously unadded) change to Accelerator_MSG Schema

tofu-rocketry commented on schemas/accelerator.sql:
```diff
@@ -0,0 +1,397 @@
+-- This schema adds the tables necessary for Accelerator accounting as a
+-- separate record as part of the wider Cloud Accounting system.
```
It's not dependant on `cloud.sql` though, is it?

RedProkofiev replied:
It is

RedProkofiev replied:
Just had a look to confirm - the only coupling is that one of the functions (SiteLookup()) is used in ReplaceAcceleratorRecord and that function is contained within cloud.sql.  I think the best approach here would be to set an issue to decouple all the miscellaneous lookup functions for categories and elements that are the same into a utils.sql, then run that inside accelerator.sql and/or cloud.sql as there's currently some duplication.  It shouldn't be a hard fix but I don't have time to figure it out

tofu-rocketry commented on requirements.txt:
```diff
@@ -4,3 +4,4 @@ MySQL-python
 python-ldap
 iso8601
 dirq
+jsonschema
```
Can you find the relevant rpm to add to the SPEC file too?

tofu-rocketry commented on apel/common/json_utils.py:
```diff
@@ -0,0 +1,29 @@
+    """ Generate JSON message format """
+    return (
+        JSON_MSG_BOILERPLATE % (type, version, list(usage_records))
+    ).replace("'", '"')
```
Needs a newline at the end of the file.

tofu-rocketry commented on apel/db/records/accelerator.py:
```diff
@@ -0,0 +1,85 @@
+
+"""
+
+from apel.common import parse_fqan
```
Unused.

tofu-rocketry commented on apel/db/records/accelerator_summary.py:
```diff
@@ -0,0 +1,84 @@
+
+"""
+
+from apel.common import parse_fqan
```
Unused.

tofu-rocketry commented on apel/common/json_utils.py:
```diff
@@ -0,0 +1,29 @@
+    "UsageRecords": %s
+}"""
+
+def to_message(type, version, *usage_records):
```
Arg name needs changing - `type` is a Python built-in.

nick-whyatt committed:
Review Changes: newlines, specfile addition, removing unused functions

tofu-rocketry committed:
Remove trailing whitespace in accelerator.sql

tofu-rocketry commented:
> this is not a fun conflict :(

I've figured out the rebase. You didn't touch anything that had actually been changed in `dev`, but the changes are just adjacent enough that they conflicted.

tofu-rocketry committed:
Code quality tweaks

- Reorder imports
- Remove unused 'as' after 'except'
- Rename method arg to snake case

tofu-rocketry committed:
Re-add separate cloud and accelerator summarising

We want to be able to keep the two types of database separate so we need
to be able to summarise cloud and accerlator databases differently.

tofu-rocketry committed:
Fix trailing whitespace

tofu-rocketry committed:
Update unittest method name which has changed

assertRaisesRegexp was added in Py 3.1. The name was changed to
assertRaisesRegex in Python 3.2.

tofu-rocketry committed:
Remove duplicated code and fix redeine of 'type'

tofu-rocketry reviewed (commented):
From Greg on #264:
> `SiteName VARCHAR(255) NOT NULL` should be `SiteID INT NOT NULL` to be consistent with other forms of APEL accounting (both in the `AcceleratorRecords` and `AcceleratorSummaries` tables) and various procedures updating accordingly.
> 
> Was the intention the accelerators would be an "add-on" to cloud or grid accounting, or its own standalone type of accounting? If it needs to be standalone - a `Sites` table and `SitesLookup` function would need to be added similar to https://github.com/apel/apel/blob/1.9.2-1/schemas/server.sql#L420

Yes, we will tweak the schema to make the accelerator accounting standalone as this will make it much easier to tweak and develop. It should also match the cloud schema a lot closer generally such as having update time columns.

tofu-rocketry commented:
Closed in favour of #459
